### ISSUE TITLE
Railtie.respond_to_missing? creates an instance unexpectedly

### ISSUE THREAD

**Issue by eregon** [rails/rails#44761]
### Steps to reproduce

The code is https://github.com/rails/rails/blob/ad25dfad8797b5024d90592bdb6a5b2d725a4850/railties/lib/rails/railtie.rb#L216-L218

This can lead to very confusing bugs, because it's not safe to e.g. check `respond_to?(:ruby2_keywords, true)` on Railtie or any of the subclasses.

So this line in Rails 6.1 actually creates an instance of `Railtie`, that's very likely unexpected:
https://github.com/rails/rails/blob/d38ff07a39c98f5db2b15ca3a51acee48c972205/railties/lib/rails/railtie.rb#L212
https://github.com/rails/rails/blob/6-1-stable/railties/lib/rails/railtie.rb

An example bug caused by this is RSpec for instance can't use `respond_to?(:ruby2_keywords, true)` when mock'ing `::Rails.application` (and so defining methods on the singleton class of the `::Rails.application`).
(details in https://github.com/rspec/rspec-mocks/pull/1385#issuecomment-755340298)

In short, this basically breaks `respond_to?` (or unintentionally creates instances of them) on Railtie and all its subclasses (on the class objects themselves), and so makes it difficult to check if e.g. some method defined is available on the class.

I think ideally Rails should never define `respond_to_missing?` on modules/classes objects themselves, it should only do so for instances of classes.

A workaround for the specific case of checking if ruby2_keywords is available is to use `Module.private_method_defined?(:ruby2_keywords)` instead.

**Comment by rafaelfranca** [comment]
I'm not sure there is any other way to do what we want. Railties is using the singleton pattern. There is only one instance and people should not know it exists. If we don't implement `respond_to_missing?` or `respond_to?` for that matter they will not be able to know the methods existent in the singleton exists or not. What is the problem of creating a new instance, just because of the mocks?

**Comment by rafaelfranca** [comment]
I actually not sure what that Rspec issue is about. Creating an instance of the Railtie only fail if you try to create an instance of an abstract raitie and the Rails application is not an abstract railtie.

```ruby
class Foo < Rails::Application
end
Foo.respond_to?(:something)
=> false
```

So if the person is mocking the `Rails.application` which is not an abstract railtie they should not see any exception, or any non-abstract railtie. Now if they are mocking `Rails::Railtie` or `Rails::Engine` directly, I'm not sure there is anything we can do.

**Comment by rafaelfranca** [comment]
I end up figuring out what was happening. Fixed.

**Comment by eregon** [comment]
Thank you for the fix! That should help avoiding the confusing case.

I think ideally the singleton pattern would be explicit (using `SomeClass.instance.foo`), that way `SomeClass` wouldn't mix the behavior of both being a class and an instance. That's likely not easily fixable for Rails if `SomeRailtie.method_of_the_instance` is used and so would be a compatibility issue.
